tribits/core/package_arch/TribitsSystemDataStructuresMacrosFunctions.rst

```diff
@@ -2,126 +2,529 @@ TriBITS System Data Structures
 
-The basic variables that define a TriBITS Project, Repository, Package and
-Subpackage are listed at the below links:
+TriBITS naming conventions
+++++++++++++++++++++++++++
+
+Before describing the TriBITS package architecture data-structures and the
+macros/functions that create and manipulate those data-structures in detail,
+first we define some naming conventions for TriBITS macros/function and
+variables.
+
+First, the term "Package" is overloaded in the context of TriBITS (and in
+CMake for that matter).  In the context of TriBITS, here are the **different
+types of entities called a "Package"**:
+
+* **TriBITS External packages/TPLs** listed in the `<repoDir>/TPLsList.cmake`_
+  file and require a `<tplDefsDir>/FindTPL<tplName>.cmake`_ file.
+
+* **TriBITS Top-level internal packages** directly listed in the
+  `<repoDir>/PackagesList.cmake`_ file
+
+* **TriBITS Subpackages** for a top-level package (defined by the
+  ``SUBPACKAGES_DIRS_CLASSIFICATIONS_OPTREQS`` argument in
+  `tribits_package_define_dependencies()`_ command called in the
+  `<packageDir>/cmake/Dependencies.cmake`_ file)
+
+* **Native CMake packages** which are found using ``find_package()`` and
+  require a ``<Package>Config.cmake`` package config file or a
+  ``Find<Package>.cmake`` find module file
+
+* **CMake Packaging** of a CMake project which means breaking it into
+  components and provide various source and binary distributions.  (A TriBITS
+  top-level package maps to a CMake packaging "component".)
+
+Also, there are the collection of all three of the TriBITS-related "packages".
+
+To try to avoid ambiguity, we define the following identifiers that appear in
+TriBITS variable, macro, and function names:
+
+* **TPL** or **EXTERNAL_PACKAGE**: Variable or function specifically for
+  processing TriITS external packages/TPLs.
+
+* **TOPLEVEL_PACKAGE**: Variable or a function specifically for a top-level
+  package (or sometimes if there is no ambiguity just **PACKAGE**).
+
+* **SUBPACKAGE**: Variable or a function specifically for a subpackage (and
+  **SUBPACKAGE_FULLNAME** if it is the full name of a subpackage which
+  includes the top-level parent package name
+  ``${PARENT_PACKAGE_NAME}${SUBPACKAGE}``).
+
+* **PACKAGE**: Variable or function that could pertain to an external
+  package/TPL, or a top-level package, or a subpackage. (The old name was
+  ``SE_PACKAGE`` in legacy TriBITS.)
+
+TriBITS uses the follow general case naming conventions for variables, macros,
+functions and module files:
+
+* ``ProperNoun_UPPER_CASE`` is generally used for global and cache variables.
+  The proper nouns using CamelCase include the names of TriBITS entities like
+  Projects, Repositories, and Packages.  UPPER_CASE is used for non-proper
+  nouns. (E.g. ``Trilinos_SOURCE_DIR``)
+
+* ``camelCaseVariableName`` or ``lower_case_variable_name`` is generally used
+  for local variable names.
+
+* ``tribits_lower_case_name()`` in generally used for TriBITS functions and
```

Comment:
> Also, did you ever decide to adopt a different naming convention, like `CamelCase_CamelCase()` for functions to make the function names more readable, like we discussed a while back?

@KyleFromKitware, not yet.  As I stated before, I am more used to functions (verbs) starting with lower-case names and proper nouns starting with a capital letter.  So it would be `ProperNoun_someFunction()`.  (That is a common convention in some other languages.)  What do you think about that?  But TriBITS would have to actually implement that before we document it.

Reply:
Fixed in new commit https://github.com/bartlettroscoe/TriBITS/commit/a7d7e43bc8012e7c195d8d5e412538e2b22e434d.